fixes for gleam 0.33

gdamjan opened the pull request:
syntax changes to update to latest Gleam
- import types with import a/b.{type C}
- pub external type -> pub opaque type
- BitString -> BitArray
- pub external fn -> @external(erlang, …

- update dependencies, especially gleeunit 0.10 was incompatible with gleam 0.33

gdamjan committed:
fixes for gleam 0.33

- update dependencies, especially gleeunit 0.10 was incompatible with
  gleam 0.33

syntax changes to update to latest Gleam
- import types with import a/b.{type C}
- pub external type -> pub opaque type
- BitString -> BitArray
- pub external fn -> @external(erlang, …

gdamjan commented:
tests don't pass though:
```
   Compiled in 0.23s
    Running nerf_test.main
F
Failures:

  1) nerf_test.echo_test
     Failure: undef
     stacktrace:
       gun.ws_send
       nerf/gun.ws_send
       nerf_test.echo_test
       nerf_test.echo_test
     output: 

Pending:
  nerf_test.echo_wss_test
    %% Unknown error: {timeout,
                   #{stacktrace =>
                         [{gun,await_up,3,
                              [{file,
                                   "/home/damjan/src/nerf/build/dev/erlang/gun/src/gun.erl"},
                               {line,838}]},
                          {gun,await_up,1,
                              [{file,
                                   "/home/damjan/src/nerf/build/dev/erlang/gun/src/gun.erl"},
                               {line,820}]},
                          {nerf@websocket,'-connect/4-fun-4-',3,
                              [{file,
                                   "/home/damjan/src/nerf/build/dev/erlang/nerf/_gleam_artefacts/nerf@websocket.erl"},
                               {line,53}]},
                          {nerf_test,echo_wss_test,0,
                              [{file,
                                   "/home/damjan/src/nerf/build/dev/erlang/nerf/_gleam_artefacts/nerf_test.erl"},
                               {line,165}]},
                          {eunit_test,'-mf_wrapper/2-fun-0-',2,
                              [{file,"eunit_test.erl"},{line,273}]},
                          {eunit_test,run_testfun,1,
                              [{file,"eunit_test.erl"},{line,71}]},
                          {eunit_proc,run_test,1,
                              [{file,"eunit_proc.erl"},{line,543}]},
                          {eunit_proc,with_timeout,3,
                              [{file,"eunit_proc.erl"},{line,368}]}]}}
  undefined
    %% Unknown error: {blame,[1,2]}


Finished in ? seconds
4 tests, 1 failures, 3 cancelled
```

gdamjan committed:
bit_builder.from_bit_string replaced with gleam/bytes_builder from_bit_array

gdamjan committed:
gun:ws_send requires the StreamRef

and so `nerf_ffi:ws_send_erl` will need it too.
this is a change from gun 2.0

so move `ws_send/2` and `ws_send_erl` in src/nerf/websocket.gleam and pass it
the whole Connection (not just ConnectionPid)

- remove websocket.Frame, there's already gun.Frame.
- fix test to use gun.Frame type

gdamjan commented:
ws:// tests now pass with the podman/docker image.

wss:// on socketsbay.com are a bit problematic, since it seems there's a lot of garbage on those sockets. looking into some wss:// echo services now.

lpil commented on src/nerf/gun.gleam:
```diff
@@ -1,47 +1,37 @@
-external fn ws_send_erl(ConnectionPid, Frame) -> OkAtom =
-  "nerf_ffi" "ws_send_erl"
-
-pub fn ws_send(pid: ConnectionPid, frame: Frame) -> Nil {
```
I think this function has been mistakenly removed?

gdamjan replied:
individual commits had a bit more context.

`ws_send/2` and `ws_send_erl` were moved in in src/nerf/websocket.gleam since they require Connection (not just ConnectionPid). 

Now another approach would be to split the types into their own file, so we can import them everywhere, or … circular imports (are they allowed?)

lpil replied:
Please revert this breaking change, thank you

lpil commented on src/nerf/websocket.gleam:
```diff
@@ -1,20 +1,14 @@
-  Close
-  Text(String)
-  Binary(BitString)
-}
```
I think this has been mistakenly removed?

gdamjan replied:
type Frame exists as `gun.Frame` already. it was not clear if it was intentional to have two separate types with the same name?

lpil replied:
The API is intentionally designed, the documentation explains the goals.

gdamjan replied:
> The API is intentionally designed, the documentation explains the goals.

maybe I'm missing something. Can you point me to the documentation so I can understand this better.

https://hexdocs.pm/nerf/ is a bit sparse

lpil replied:
They are deliberately different. The nerf module is a low level binding.

lpil reviewed (commented):
Hello! The upgrade looks good but please revert the changes to the API. Thank you

gdamjan committed:
gleam/bit_builder is removed in stdlib v0.35.0

replaced with gleam/bytes_builder

https://github.com/gleam-lang/stdlib/blob/v0.35.0/CHANGELOG.md?plain=1#L10

gdamjan committed:
update-deps

jluzny commented:
I tested the latest upgrade in my [fork of glome](https://github.com/jluzny/glome/tree/refactor/migration-to-gleam-1.0), and it seems to be working just fine :)

lpil commented:
Thanks @jluzny but this is PR is unfinished presently.
